sopel/modules/currency.py

```diff
@@ -2,109 +2,201 @@
 """
 currency.py - Sopel Currency Conversion Module
 Copyright 2013, Elsie Powell, embolalia.com
+Copyright 2019, Mikkel Jeppesen
 Licensed under the Eiffel Forum License 2.
 
 https://sopel.chat
 """
 from __future__ import unicode_literals, absolute_import, print_function, division
 
 import re
+import time
 
-from requests import get
+import requests
 
-from sopel.module import commands, example, NOLIMIT
+from sopel.config.types import StaticSection, ValidatedAttribute
+from sopel.logger import get_logger
+from sopel.module import commands, example, NOLIMIT, rule
 
 
-# The Canadian central bank has better exchange rate data than the Fed, the
-# Bank of England, or the European Central Bank. Who knew?
-base_url = 'https://www.bankofcanada.ca/valet/observations/FX{}CAD/json'
-regex = re.compile(r'''
-    (\d+(?:\.\d+)?)        # Decimal number
-    \s*([a-zA-Z]{3})       # 3-letter currency code
-    \s+(?:in|as|of|to)\s+  # preposition
-    ([a-zA-Z]{3})          # 3-letter currency code
-    ''', re.VERBOSE)
+FIAT_URL = 'https://api.exchangeratesapi.io/latest?base=EUR'
+FIXER_URL = 'http://data.fixer.io/api/latest?base=EUR&access_key={}'
+CRYPTO_URL = 'https://apiv2.bitcoinaverage.com/indices/global/ticker/short?crypto=BTC'
+EXCHANGE_REGEX = re.compile(r'''
+    ^(\d+(?:\.\d+)?)                                            # Decimal number
+    \s*([a-zA-Z]{3})                                            # 3-letter currency code
+    \s+(?:in|as|of|to)\s+                                       # preposition
+    (([a-zA-Z]{3}$)|([a-zA-Z]{3})\s)+$                          # one or more 3-letter currency code
+''', re.VERBOSE)
+LOGGER = get_logger(__name__)
+UNSUPPORTED_CURRENCY = "Sorry, {} isn't currently supported."
+UNRECOGNIZED_INPUT = "Sorry, I didn't understand the input."
 
+rates_fiat_json = {}
```

Comment:
Initialize with the expected structure to prevent the issues I pointed out (`if of in rates_fiat_json['rates']...`). Just a suggestion:
```suggestion
rates_fiat_json = {'rates':{}}
```

**OR**

Change the conditions below to first check if `'rates'` is in this dictionary... this may be the better option since then you can also detect errors.